This is the first guided project that we are going to implement. Again here we use *autos.csv* dataset. Where some of the column name are dataCrawled, name, seller, offerType, price, abtest, vehicleType etc.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [29]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType  price   abtest vehicleType  registration_year  \
0      privat   Angebot   5000  control         bus               2004   
1      privat   Angebot   8500  control   limousine               1997   
2      privat   Angebot   8990     test   limousine               2009   
3      privat   Angebot   4350  control  kleinwagen               2007   
4      privat   Angebot   1350     test       kombi               2003   
...       ...       ...    ...      ...         ...                ...   
49995  privat   Angebot  24900  control   limousine               2011   
49996  privat   Angebot   1980  control      cabrio               1996   
49997  privat   Angebot  13200     test      cabrio               2014   
49998  privat   Angebot  22900  control       kombi               2013   
49999  privat   Angebot   1250  control   limousine               1996   

         gearbox  powerPS   model  odometer_km  registration_month fuelType  \
0        manuell      158  andere       150000                   3      lpg   
1      automatik      286     7er       150000                   6   benzin   
2        manuell      102    golf        70000                   7   benzin   
3      automatik       71  fortwo        70000                   6   benzin   
4        manuell        0   focus       150000                   7   benzin   
...          ...      ...     ...          ...                 ...      ...   
49995  automatik      239      q5       100000                   1   diesel   
49996    manuell       75   astra       150000                   5   benzin   
49997  automatik       69     500         5000                  11   benzin   
49998    manuell      150      a3        40000                  11   diesel   
49999    manuell      101  vectra       150000                   1   benzin   

            brand  unrepaired_damage           ad_created  nrOfPictures  \
0         peugeot               nein  2016-03-26 00:00:00             0   
1             bmw               nein  2016-04-04 00:00:00             0   
2      volkswagen               nein  2016-03-26 00:00:00             0   
3           smart               nein  2016-03-12 00:00:00             0   
4            ford               nein  2016-04-01 00:00:00             0   
...           ...                ...                  ...           ...   
49995        audi               nein  2016-03-27 00:00:00             0   
49996        opel               nein  2016-03-28 00:00:00             0   
49997        fiat               nein  2016-04-02 00:00:00             0   
49998        audi               nein  2016-03-08 00:00:00             0   
49999        opel               nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
49995       8213

Here DataFrame.head() print or indicate first some rows from the *autos.csv* file. Here we print 5 rows.

In [4]:
#df = pd.DataFrame()
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Here DataFrame.info() method give us the total overview about our dataset *autos.csv*. 

In [5]:
autos.info(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [6]:
col = list(autos.columns)
print(col)

['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest', 'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model', 'odometer', 'monthOfRegistration', 'fuelType', 'brand', 'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode', 'lastSeen']


Here in this section we try to find out the columns name using 
DataFrame.columns attribute. Then we change some of the columns name such as *yearOfRegistration* to *registration_year* etc. After that we assigned the changed columns name to the original list. Using DataFrame.head() attribute we examined the changed stuff.

In [7]:
i = -1
for row in col:
    i += 1
    if row == 'yearOfRegistration':
        col[i] = 'registration_year'
    elif row == 'monthOfRegistration':
        col[i] = 'registration_month'
    elif row == 'notRepairedDamage':
        col[i] = ' unrepaired_damage'
    elif row == 'dateCreated':
        col[i] = 'ad_created'
    
print(col)        

['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest', 'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model', 'odometer', 'registration_month', 'fuelType', 'brand', ' unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode', 'lastSeen']


In [8]:
autos.columns = col
print(list(autos.columns))

['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest', 'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model', 'odometer', 'registration_month', 'fuelType', 'brand', ' unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode', 'lastSeen']


In [9]:
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuelType  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand  unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot               nein  2016-03-26 00:00:00             0   
1         bmw               nein  2016-04-04 00:00:00             0   
2  volkswagen               nein  2016-03-26 00:00:00             0   
3       smart               nein  2016-03-12 00:00:00             0   
4        ford               nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Here *DataFrame.describe()* method is used to view some basic statistical details like percentile, mean, std etc. of a data frame.

In [10]:
autos.describe(include = 'all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  registration_year  gearbox       powerPS  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuelType       brand  unrepaired_damage  \
count       50000        50000.000000    45518       50000              40171   
unique         13                 NaN        7          40                  2   
top     150,000km                 NaN   benzin  volkswagen               nein   
freq        32424                 NaN    30107       10687              35232   
mean          NaN            5.723360      NaN         NaN                NaN   
std           NaN            3.711984      NaN         NaN                NaN   
min           NaN            0.000000      NaN         NaN                NaN   
25%           NaN            3.000000      NaN         NaN                NaN   
50%           NaN            6.000000      NaN         NaN                NaN   
75%           NaN            9.000000      NaN         NaN                NaN   
max           NaN           12.000000      NaN         NaN                NaN   

                 ad_created  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

Removing any non-numeric characters from *price* column

In [11]:
# for row in autos[][:

#     row = re.sub("[^0-9]", "", row)

p = list(autos['price'])

for index,row in enumerate(p):
   
    numeric_filter = filter(str.isdigit, row)
    numeric_string = "".join(numeric_filter)

    row = numeric_string
   
    p[index] = row
autos['price'] = p
print(autos['price'])




0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: object


Removing any non-numeric character from *odometer* column

In [12]:
q = list(autos['odometer'])

for index,row in enumerate(q):
   
    numeric_filter = filter(str.isdigit, row)
    numeric_string = "".join(numeric_filter)

    row = numeric_string
   
    q[index] = row
autos['odometer'] = q
print(autos['odometer'])



0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 50000, dtype: object


Converting the column to a numeric dtype

In [13]:
autos['price'] = autos['price'].apply(pd.to_numeric)
# autos['price'] = autos['price'].astype(int)


In [14]:
autos['odometer'] = autos['odometer'].apply(pd.to_numeric)

Renaming the *odometer* to *odometer_km*

In [15]:
autos.rename(columns = {'odometer' : 'odometer_km'}, inplace = True)
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType  price   abtest vehicleType  registration_year    gearbox  \
0  privat   Angebot   5000  control         bus               2004    manuell   
1  privat   Angebot   8500  control   limousine               1997  automatik   
2  privat   Angebot   8990     test   limousine               2009    manuell   
3  privat   Angebot   4350  control  kleinwagen               2007  automatik   
4  privat   Angebot   1350     test       kombi               2003    manuell   

   powerPS   model  odometer_km  registration_month fuelType       brand  \
0      158  andere       150000                   3      lpg     peugeot   
1      286     7er       150000                   6   benzin         bmw   
2      102    golf        70000                   7   benzin  volkswagen   
3       71  fortwo        70000                   6   benzin       smart   
4        0   focus       150000                   7   benzin        ford   

   unrepaired_damage           ad_created  nrOfPictures  postalCode  \
0               nein  2016-03-26 00:00:00             0       79588   
1               nein  2016-04-04 00:00:00             0       71034   
2               nein  2016-03-26 00:00:00             0       35394   
3               nein  2016-03-12 00:00:00             0       33729   
4               nein  2016-04-01 00:00:00             0       39218   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [16]:
max_price = max(autos['price'])
print(max_price)
min_price = min(autos['price'])
print(min_price)

99999999
0


In [17]:
max_odo = max(autos['odometer_km'])
print(max_odo)
min_odo = min(autos['odometer_km'])
print(min_odo)

150000
5000


In [18]:
autos['price'].unique().shape

(2357,)

In [19]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [20]:
autos['price'].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

In [21]:
autos['price'].sort_index(ascending = True).head(50)

0      5000
1      8500
2      8990
3      4350
4      1350
5      7900
6       300
7      1990
8       250
9       590
10      999
11      350
12     5299
13     1350
14     3999
15    18900
16      350
17     5500
18      300
19     4150
20     3500
21    41500
22    25450
23     7999
24    48500
25       90
26      777
27        0
28     5250
29     4999
30       80
31     2850
32      350
33     4800
34     2600
35     1200
36     7500
37    13800
38     2850
39     1450
40     3500
41     3200
42    11900
43    11500
44      900
45     1000
46     9000
47     3499
48     1099
49    56000
Name: price, dtype: int64

2.1.7.5:  Exploring the Date columns. *dateCrawled*

In [26]:
autos['dateCrawled'].value_counts(normalize = True, dropna = False)

2016-04-02 11:37:04    0.00006
2016-04-04 16:40:33    0.00006
2016-03-21 20:37:19    0.00006
2016-03-09 11:54:38    0.00006
2016-04-02 15:49:30    0.00006
                        ...   
2016-03-26 12:47:05    0.00002
2016-03-05 21:51:43    0.00002
2016-03-25 10:38:50    0.00002
2016-03-12 15:51:30    0.00002
2016-03-20 16:38:58    0.00002
Name: dateCrawled, Length: 48213, dtype: float64

In [28]:
autos['dateCrawled'].sort_index(ascending =True)

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
                ...         
49995    2016-03-27 14:38:19
49996    2016-03-28 10:50:25
49997    2016-04-02 14:44:48
49998    2016-03-08 19:25:42
49999    2016-03-14 00:42:12
Name: dateCrawled, Length: 50000, dtype: object

2.1.7.5:   Exploring the date columns *last_seen*

In [31]:
autos['ad_created'].value_counts(normalize = True, dropna =False)

2016-04-03 00:00:00    0.03892
2016-03-20 00:00:00    0.03786
2016-03-21 00:00:00    0.03772
2016-04-04 00:00:00    0.03688
2016-03-12 00:00:00    0.03662
                        ...   
2015-12-30 00:00:00    0.00002
2016-01-13 00:00:00    0.00002
2015-08-10 00:00:00    0.00002
2016-01-03 00:00:00    0.00002
2016-01-29 00:00:00    0.00002
Name: ad_created, Length: 76, dtype: float64

In [32]:
autos['ad_created'].sort_index(ascending = True)

0        2016-03-26 00:00:00
1        2016-04-04 00:00:00
2        2016-03-26 00:00:00
3        2016-03-12 00:00:00
4        2016-04-01 00:00:00
                ...         
49995    2016-03-27 00:00:00
49996    2016-03-28 00:00:00
49997    2016-04-02 00:00:00
49998    2016-03-08 00:00:00
49999    2016-03-13 00:00:00
Name: ad_created, Length: 50000, dtype: object

2.1.7.5: Exploring the date columns *last_seen*

In [34]:
autos['lastSeen'].value_counts(normalize = True, dropna = False)

2016-04-07 06:17:27    0.00016
2016-04-06 21:17:51    0.00014
2016-04-06 06:17:24    0.00014
2016-04-07 03:16:17    0.00014
2016-04-06 02:16:12    0.00012
                        ...   
2016-03-10 08:46:32    0.00002
2016-03-26 19:45:41    0.00002
2016-04-06 14:47:11    0.00002
2016-03-11 17:52:35    0.00002
2016-03-31 02:45:45    0.00002
Name: lastSeen, Length: 39481, dtype: float64

In [ ]:
autos['lastSeen'].sort_index(ascending = True)

In [36]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Here in the *registration_year* column we see that the maximum value or year is 9999 and the min value or year is 1000. Which seems to look impractical. Because 9999's is too far from the present day. And again in 1000's car was not invented.

2.1.7.6: Dealing with Incorrect Registration year Data

In [46]:
ots = []
for row in autos['registration_year']:
    if row > 2016 or row <1900:
        ots.append(row)

print(ots)        
# num_ots = ots.count()
print(len(ots))

[2017, 2017, 2017, 2017, 2018, 2017, 2018, 2017, 2017, 2017, 2018, 2017, 4500, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2018, 2017, 2017, 2017, 2018, 2017, 2017, 2017, 2017, 2018, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2018, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2018, 2018, 2017, 2017, 2017, 2017, 2017, 2018, 2017, 2018, 2017, 2018, 2018, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2018, 2017, 2017, 2017, 2018, 2017, 2018, 2017, 2017, 2017, 2017, 2018, 2017, 2017, 2017, 2018, 2017, 2017, 2017, 2018, 2018, 2018, 2017, 2017, 2017, 2017, 2018, 2018, 2017, 2017, 2017, 2018, 2018, 2017, 2017, 2017, 2017, 2017, 2018, 2017, 2017, 2017, 2018, 2017, 2018, 2017, 2018, 2017, 2017, 2017, 2017, 2017, 2017, 2017, 2018, 2017, 2018, 2017, 5000, 2017, 2017, 2018, 2018, 2017, 2017, 2017, 2017, 2017, 2018, 2017, 2017, 2018, 2018, 2017, 410

Here from above column we see that there are **1972** columns or car that fall outside the **1900-2016** interval.

2.1.7.7: Exploring Price by Brand

In [50]:
autos['brand'].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
citroen             701
smart               701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

In [54]:
vc = autos['brand'].value_counts(normalize = True)
print(vc)

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
citroen           0.01402
smart             0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

In [58]:
# top_20 = []
# for row in vc:
#     if row > .05:
#         top_20.append(row)
# print(top_20)        
top_20 = vc > 0.05
print(top_20)

volkswagen         True
opel               True
bmw                True
mercedes_benz      True
audi               True
ford               True
renault           False
peugeot           False
fiat              False
seat              False
skoda             False
mazda             False
nissan            False
citroen           False
smart             False
toyota            False
sonstige_autos    False
hyundai           False
volvo             False
mini              False
mitsubishi        False
honda             False
kia               False
alfa_romeo        False
porsche           False
suzuki            False
chevrolet         False
chrysler          False
dacia             False
daihatsu          False
jeep              False
subaru            False
land_rover        False
saab              False
daewoo            False
trabant           False
jaguar            False
rover             False
lancia            False
lada              False
Name: brand, dtype: bool


In [63]:
per_5 = vc[top_20]
print(per_5)

volkswagen       0.21374
opel             0.10922
bmw              0.10858
mercedes_benz    0.09468
audi             0.08566
ford             0.06958
Name: brand, dtype: float64


Here from above we see that total **six car band** have over a percentage over 5%.


volkswagen       0.21374
opel             0.10922
bmw              0.10858
mercedes_benz    0.09468
audi             0.08566
ford             0.06958

In [65]:
agg_data = {}
for row in top_20:
    per_5 = vc[top_20]
    if per_5 in agg_data:
        agg_data[per_5]
    
        

TypeError: 'Series' objects are mutable, thus they cannot be hashed

2.1.7.9: 